### ISSUE TITLE
iHPC apps should support rendering erb files in the template directory

### ISSUE THREAD

**Issue by ericfranz** [OSC/ood-dashboard#179]
Take https://github.com/OSC/bc_osc_jupyter/tree/01f68ce8253a86a47842c01a7e6d14ec86b1a9d0 as an example.

Lets say I wanted to add a checkbox to the webform "Use rmpi" and when checked the script.sh will do `module load openmpi`.

It appears that the only option I have to implement this is currently to do this.

1. modify form.yml.erb

```diff
#...
attributes:
+  user_rmpi: { widget: checkbox, label: "Use rmpi" }
  #...
form:
  - bc_num_hours
  - bc_num_slots
  - node_type
  - bc_account
  - bc_email_on_started
+ - user_rmpi
```

2. modify submit.yml.erb

```diff
---
script:
  job_environment:
    PYTHON_MODULE: "python/3.5"
    <%- if node_type.include? "gpus" -%>
    CUDA_MODULE: "cuda/8.0.44"
    <%- end -%>
+   <%= "OPENMPI_MODULE: 'openmpi'" if use_rmpi %>
```

3. modify template/script.sh

```diff
 # Load the require modules
 module load ${PYTHON_MODULE}
 [[ ${CUDA_MODULE} ]] && module load ${CUDA_MODULE}
+[[ ${OPENMPI_MODULE} ]] && module load ${OPENMPI_MODULE}
```

The use of the intermediate env variable in order to modify the script seems limiting. With our app kit apps the entire job template can contain template files which are rendered. If we supported this same approach then I could skip the extra USE_RMPI env var:

2. modify template/script.sh.erb

```diff
 # Load the require modules
 module load ${PYTHON_MODULE}
 [[ ${CUDA_MODULE} ]] && module load ${CUDA_MODULE}
+<%= "module load openmpi" if use_rmpi %>
```

That said, this is currently the only example in iHPC apps where we are using an env var to compensate for the lack of templating in the job scripts. So we can hold off until we have more use cases demonstrated here.

**Comment by ericfranz** [comment]
@nickjer can you provide details on why this change would be difficult with links to relevant source code?

**Comment by nickjer** [comment]
Batch connect is designed to work with two distinct directories in the Batch Connect App's defined dataroot.

- `output/`
- `staged/`

The `output/` directory is where all the job specific data lies:

- any app-specific configuration files that hold job-specific data
- password file
- connection information file
- log files

So all of these files reside in job-specific sub-directories underneath the `output/` root directory. This keeps job-specific files isolated from each other.

The `staged/` directory contains all the files/assets common among every job:

- the scripts run to launch the GUIs and/or web servers
- configuration files for window managers
- icons or image files used by app
- basically these are all the files that reside under the `template/` directory of each Batch Connect app

These files should **never** change between jobs. This allows us to reduce the disk space used after a user has launched thousands of jobs over time (especially if the app comes with large assets, e.g., images or binaries).

By allowing the files under the Batch Connect App's `template/` directory to be modified for each job, you are in effect removing the need for the dataroot's `staged/` directory and copying all `template/` files to the dataroot's `output/<job_uuid>/` directory.

**Comment by ericfranz** [comment]
What is the largest size staged directory we have seen?

Hypothetical: If we had a custom staged directory for each job i.e. `staged/<job_uuid>`, and we deleted the directory when we delete the db record file, then we would only have multiple copies in the user's home directory if they were running multiple sessions of the same app.

**Comment by nickjer** [comment]
When we included only the 64 bit version of `proot` with the RStudio server app (before we asked Systems to install it locally on the Owens cluster), it would weigh in over 6 MB.

**Comment by nickjer** [comment]
When we used FVWM as the window manager, the apps could weigh as much as ~110 KB.

**Comment by nickjer** [comment]
So only issue would be if the template included `*.erb` files meant to be processed within the batch job. But the user could avoid this by renaming the extension as `*.erb.txt` or some other extension. Then have his batch job process that file directly as ERB.

**Comment by nickjer** [comment]
@ericfranz 

Another issue is which object `Session` or `SessionContext` to bind the ERB with. `Session` holds important information like `#output_root` and `#staged_root`. While `SessionContext` holds all the attributes and their values.

Maybe we just go the route of binding both with an `OpenStruct` that combines the two as:

```sh
#!/bin/bash -l

# Set working directory to home directory
cd "${HOME}"

#
# Start RStudio Server
#

# Restore the module environment to avoid conflicts
module restore

# Load the require modules
module load <%= context.r_module %>
module load <%= context.rstudio_module %>

# Launch the RStudio Server
echo "Starting up rserver..."
<%= context.proot_path %> -b ${TMPDIR}:/tmp \
  rserver \
    --www-port ${port} \
    --auth-none 0 \
    --auth-pam-helper-path "<%= session.staged_root.join("bin", "auth") %>" \
    --auth-encrypt-password 0
```

**Comment by ericfranz** [comment]
>`Session` holds important information like `#output_root` and `#staged_root`

This currently is not used in the template as it is. So how would having access to this change things? Are you imagining an alternative to using PBS_O_WORKDIR?

Also, with the ability to use template files, we could more easily abstract all torque or OSC specific things to "configuration options" i.e. SessionContext attributes.

**Comment by nickjer** [comment]
The example above gives:

```shell
    --auth-pam-helper-path "<%= session.staged_root.join("bin", "auth") %>" \
```

> This currently is not used in the template as it is.

It is determined within the template using some Bash-fu:

```shell
# Get source directory of this script
export STAGED_ROOT="$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")"
```

but would be ideal if it could be supplied within ERB, similar to how we are going to supply the context parameters.

Also, this leads to even crazier Bash-fu in the case of the Ansys app with:

- https://github.com/OSC/bc_osc_ansys_workbench/blob/master/template/script.sh#L41
- https://github.com/OSC/bc_osc_ansys_workbench/blob/master/template/fluxbox/menu#L6

**Comment by ericfranz** [comment]
Remind me why `#output_root` and `#staged_root` will be required as they both will be the same?

**Comment by ericfranz** [comment]
One thought: the nice thing we do get from having a separate staged directory step (even if there is a separate staged directory per job) is that:

1. the files created by the job are in a separate directory from the original job template
2. the name of the directory created after the job begins is the name of the job id

With the batch sim demo you put together a while back, you had similar separation, with a dedicated results directory, which had the use of being able to auto-load results.

So even if we have one directory that is staged to and that is the job directory, maybe during the execution of the job, directories we create to add certain files can follow a certain convention we agree on so we get some of these benefits but use the same (or very similar model) of managing job data between iHPC, batch sim, appkit gateway apps, etc.

**Comment by nickjer** [comment]
The pull request #211 already removes the directory separation. And it may take too much time right now to work on a new directory structure design.
